lib/src/idx/planner/tree.rs

```diff
@@ -0,0 +1,162 @@
+use crate::dbs::{Options, Transaction};
+use crate::err::Error;
+use crate::idx::planner::plan::IndexOption;
+use crate::sql::statements::DefineIndexStatement;
+use crate::sql::{Cond, Expression, Idiom, Operator, Subquery, Table, Value};
+use async_recursion::async_recursion;
+use std::collections::hash_map::Entry;
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+pub(super) struct Tree {}
+
+pub(super) type IndexMap = HashMap<Expression, HashSet<IndexOption>>;
+
+impl Tree {
+	pub(super) async fn build<'a>(
+		opt: &'a Options,
+		txn: &'a Transaction,
+		table: &'a Table,
+		cond: &Option<Cond>,
+	) -> Result<Option<(Node, IndexMap)>, Error> {
+		let mut b = TreeBuilder {
+			opt,
+			txn,
+			table,
+			indexes: None,
+			index_map: IndexMap::default(),
+		};
+		let mut res = None;
+		if let Some(cond) = cond {
+			res = Some((b.eval_value(&cond.0).await?, b.index_map));
+		}
+		Ok(res)
+	}
+}
+
+struct TreeBuilder<'a> {
```

Comment:
I don't know if we should be calling this a Builder if it isn't actually a Builder. Im not quite sure what I would call it, as it seems to be similar behaviour to statements, without actually being a statement. Clustering has the same issues, so would be interested in the decision.

Reply:
This is quite a hidden feature! But in fact, it does build a tree. The function `eval_value` returns a `Node` which is the root of the tree.
https://github.com/surrealdb/surrealdb/blob/ad9d86cc930228cca1dacced2cd46c6ddfeedf89/lib/src/idx/planner/tree.rs#L69
But indeed, there is no "Tree" structure created. I also didn't want to call it NodeBuilder, which would have been too restrictive.

Reply:
Perhaps it could be called TreeFactory then - quite the java-ism 😅 The gripe really isn't about tree, it's the I would expect a builder to be
```
let a = SomeBuilder::new(required, parameters, here)
   .withOne(1)
   .withAnother(2)
   .withOneMore(3)
   .build();
```

Reluctant to call it factory as well, since that is more of a `TreeFactory::forSelectPlanner(selectStatement)`, which this isn't quite doing... 🤔